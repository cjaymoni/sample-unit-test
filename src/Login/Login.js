import React, { useState } from "react";
import axios from "axios";

export default function Login() {
  const [formValues, setFormValues] = useState({});
  const [loading, setLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState(false);
  const [loggedInUser, setLoggedInUser] = useState({});

  const handleInputChange = ({ target: { name, value } }) => {
    setFormValues({ ...formValues, [name]: value });
  };
  const handleSubmit = async (e) => {
    setLoading(true);
    e.preventDefault();
    try {
      const { data } = await axios.get(
        "https://jsonplaceholder.typicode.com/users/1"
      );
      setLoggedInUser(data);
    } catch {
      setErrorMsg(true);
    }
    setLoading(false);
  };

  return (
    <div className="body-class">
      <h1>Simple Login Form</h1>

      <div className="form-class">
        <form>
          <div className="form-input">
            <label htmlFor="username">Username</label>
            <input
              type="text"
              data-testid="username"
              name="username"
              placeholder="Username"
              onChange={handleInputChange}
            />
          </div>
          <div className="form-input">
            <label htmlFor="password">Password</label>
            <input
              data-testid="password"
              type="password"
              name="password"
              placeholder="Password"
              onChange={handleInputChange}
            />
          </div>
          <div>
            <button
              type="submit"
              onClick={handleSubmit}
              disabled={!formValues.username || !formValues.password}
            >
              {" "}
              {loading ? "Please hold on while fetching" : "Login"}
            </button>
          </div>
        </form>
        {errorMsg && <p>An Error Occurred</p>}
        <div className="success" data-testid="user">
          {JSON.stringify(loggedInUser)}
        </div>
      </div>
    </div>
  );
}
