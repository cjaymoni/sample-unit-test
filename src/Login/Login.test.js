import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";

import Login from "./Login";

jest.mock("axios", () => ({
  __esModule: true,
  default: {
    get: () => ({
      data: { id: 1, name: "John" },
    }),
  },
}));
describe("Login", () => {
  test("renders Login component", () => {
    render(<Login />);
    const inputText = screen.getByText("Username");
    expect(inputText).toBeInTheDocument();
    const inputPassword = screen.getByText("Password");
    expect(inputPassword).toBeInTheDocument();
  });

  test("username input should be rendered", () => {
    render(<Login />);
    const usernameNode = screen.getByPlaceholderText(/username/i);
    expect(usernameNode).toBeInTheDocument();
    expect(usernameNode.value).toBe("");
  });

  test("password input should be rendered", () => {
    render(<Login />);
    const passwordNode = screen.getByPlaceholderText(/password/i);
    expect(passwordNode).toBeInTheDocument();
    expect(passwordNode.value).toBe("");
  });

  test("username input should receive input", () => {
    render(<Login />);
    let inputValue = "hello";
    const usernameNode = screen.getByPlaceholderText(/username/i);
    fireEvent.change(usernameNode, { target: { value: inputValue } });
    expect(usernameNode.value).toBe(inputValue);
  });

  test("password input should receive input", () => {
    render(<Login />);
    let inputValue = "password";
    const passwordNode = screen.getByPlaceholderText(/password/i);
    fireEvent.change(passwordNode, { target: { value: inputValue } });
    expect(passwordNode.value).toBe(inputValue);
  });

  test("login button should not  be in a loading state", () => {
    render(<Login />);
    const loginButton = screen.getByRole("button", { name: /login/i });
    expect(loginButton).not.toHaveTextContent(/please wait/i);
  });

  test("login button should be enabled when there is input", () => {
    render(<Login />);
    const loginButton = screen.getByRole("button", { name: /login/i });

    let inputValue = "hello";
    let inputPass = "password";
    const usernameNode = screen.getByPlaceholderText(/username/i);
    const passwordNode = screen.getByPlaceholderText(/password/i);

    fireEvent.change(usernameNode, { target: { value: inputValue } });
    expect(usernameNode.value).toBe(inputValue);
    fireEvent.change(passwordNode, { target: { value: inputPass } });

    expect(passwordNode.value).toBe(inputPass);
    expect(loginButton).toBeEnabled();
  });

  test("state should be in loading on click", () => {
    render(<Login />);
    const loginButton = screen.getByRole("button", { name: /login/i });

    let inputValue = "hello";
    let inputPass = "password";
    const usernameNode = screen.getByPlaceholderText(/username/i);
    const passwordNode = screen.getByPlaceholderText(/password/i);

    fireEvent.change(usernameNode, { target: { value: inputValue } });
    expect(usernameNode.value).toBe(inputValue);
    fireEvent.change(passwordNode, { target: { value: inputPass } });
    fireEvent.click(loginButton);

    expect(loginButton).toHaveTextContent(/Please hold on while fetching/i);
  });

  test("user should be rendered in view after fetching", async () => {
    render(<Login />);
    const loginButton = screen.getByRole("button", { name: /login/i });

    let textValue = "jude";
    let textValuePass = "pass";
    const usernameNode = screen.getByPlaceholderText(/username/i);
    const passwordNode = screen.getByPlaceholderText(/password/i);

    fireEvent.change(usernameNode, { target: { value: textValue } });
    expect(usernameNode.value).toBe(textValue);
    fireEvent.change(passwordNode, { target: { value: textValuePass } });
    fireEvent.click(loginButton);
    const userItem = await screen.findByTestId("user");

    expect(userItem).toBeInTheDocument();
  });

  test("This test will fail", (done) => {
    done.fail(new Error("This is the last test supposed to fail"));
  });
});
