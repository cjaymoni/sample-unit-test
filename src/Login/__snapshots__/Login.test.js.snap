// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<div>
  <div
    class="body-class"
  >
    <h1>
      Simple Login Form
    </h1>
    <div
      class="form-class"
    >
      <form>
        <div
          class="form-input"
        >
          <label>
            Username
          </label>
          <input
            name="username"
            placeholder="Username"
            type="text"
          />
        </div>
        <div
          class="form-input"
        >
          <label>
            Password
          </label>
          <input
            name="password"
            placeholder="Password"
            type="password"
          />
        </div>
        <div>
          <button
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
